README.md
# MobileId::Ruby::Gem

Estonia Mobile ID authentication, more info at https://www.id.ee/en/

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'mobile_id'
```

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install mobile_id

## Test usage

Execute irb:

    $ bundle exec irb
    irb(main):001:0> require 'mobile_id'
    => true
    irb(main):002:0> @mid = MobileId::Auth.new(live: false)
    => #<MobileId::Auth:0x000055ac4cb25630 @url="https://tsp.demo.sk.ee/mid-api", @uuid="00000000-0000-0000-0000-000000000000", @name="DEMO", @hash="two+e7UMoFCAXHo8q9AnWqSC58Hhil74RowY8Gg9xQY=">
    irb(main):003:0> auth = @mid.authenticate!(phone: '00000766', personal_code: '60001019906')
    => {"session_id"=>"34e7eff0-691b-4fad-9798-8db680587b18", "phone"=>"00000766", "phone_calling_code"=>"+372"}
    irb(main):004:0> verify = @mid.verify!(auth)
    => {"personal_code"=>"60001019906", "first_name"=>"MARY ÄNN", "last_name"=>"O’CONNEŽ-ŠUSLIK TESTNUMBER", "phone"=>"00000766", "phone_calling_code"=>"+372", "auth_provider"=>"mobileid"}

You get verified attributes: personal_code, first_name, last_name, phone, phone_calling_code, auth_provider


## Live usage

For live usage, add your relyingPartyUUID (RPUUID) and relyingPartyName what you get from https://www.sk.ee

```ruby
    @mid = MobileId::Auth.new(live: true, uuid: "39e7eff0-241b-4fad-2798-9db680587b20", name: 'My service name')
```

Rails with Devise example controller:

```ruby
class MobileSessionsController < ApplicationController
  include Devise::Controllers::Helpers
  skip_authorization_check only: [:new, :show, :create, :update]
  before_action :init_mobile_id, only: [:create, :update]
  
  def new
    @user = User.new
  end

  # makes Mobile ID authentication
  def create
    session[:auth] = @mid.authenticate!(phone: params[:phone], personal_code: params[:personal_code])
    render :show, locals: { verification_code: @mid.verification_code }
  rescue MobileId::Error => e
    render :error, locals: { message: e }
  end

  # verifices Mobile ID user
  def update
    session[:auth] = @mid.verify!(session[:auth])
    find_or_create_user_and_redirect(personal_code: @mid.personal_code)
  rescue MobileId::Error => e
    render :error, locals: { message: e }
  end

  private

  def init_mobile_id
    @mid = MobileId.new(live: true, uuid: ENV['sk_mid_uuid'], name: ENV['sk_mid_name'])
  end

  # It's pure your system business logic what to do here with validated user attributes, example code:
  def find_or_create_user_and_redirect(personal_code:)
    @user = User.find_by(personal_code: personal_code)

    # bind currently present email only account with mobile-id
    if @user.nil? && current_user&.confirmed_email_only_account?
      @user = current_user 
      @user.personal_code = personal_code
      @user.save!
    end

    return redirect_to new_omniuser_url, notice: t(:finish_setup) if @user.nil? || @user.new_record?
    return redirect_to root_url, alert: t(:unlock_info) if @user.access_locked?

    if @user.valid? && @user.confirmed_at
      # overwrite name changes
      @user.first_name = session[:auth]['first_name'] if session[:auth]['first_name'].present?
      @user.last_name  = session[:auth]['last_name']  if session[:auth]['last_name'].present?
      @user.save! if @user.changed?

      sign_in_and_redirect(@user, notice: t('devise.sessions.signed_in'))
    else
      redirect_to edit_omniuser_url(@user), notice: t('devise.failure.unconfirmed')
    end
  end
end
```

## Development

After checking out the repo, run `bundle` to install dependencies. For testing code, run `rspec`

## Contributors

* Priit Tark
* Andri Möll for pointing out user signature issue and cert date check
* Juri Linkov for pointing out unpack method issue and test friendly init
* Thales Ribeiro for rubocop initial
* Mijora for LT locale

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/domify/mobile_id

## Roadmap

* Document sign
* Rails generators

## License

The gem is available as open source under the terms of the [MIT License](http://opensource.org/licenses/MIT).

## Sponsors

Gem development and testing is sponsored by [Domify](https://domify.io).


CHANGELOG.md
Release 0.0.15
* Restored older Ruby 2.7 support

Release 0.0.14
* Added LT locale with locale init update (thanks Mijora)
* Init rubocop (thanks Thales Ribeiro)

Release 0.0.13
* Gems updated to latest
* More test friendly init (thanks Juri Linkov)
* More precise sert check (thanks Andri Möll)

Release 0.0.12
* Supports session RUNNING state
* Validate auth signature

Release 0.0.11
* Add more live SK certs

Release 0.0.10
* Allow live certs verification in test env

Release 0.0.9
* Fixed verification code

Release 0.0.8
* Release cleanup

Release 0.0.7
* Cert cleanup

Release 0.0.6
* Cert path fix

Release 0.0.5
* Added user certificate validation 

Release 0.0.4
* Refactored MobileId to MobileId::Auth

Release 0.0.3
* Gemspec update

Release 0.0.2
* Readme update

Release 0.0.1
* Init


spec/spec_helper.rb
# frozen_string_literal: true

require 'bundler/setup'
Bundler.setup

require 'pry'
require 'mobile_id'
require 'i18n'
I18n.config.available_locales = :en
I18n.load_path << Dir["#{File.expand_path('lib/mobile_id/locales')}/*.yml"]

RSpec.configure do |config|
  config.expect_with(:rspec) { |c| c.syntax = :should }
end


spec/mobile_id_spec.rb
# frozen_string_literal: true

require 'spec_helper'

describe MobileId do
  before do
    @mid = MobileId::Auth.new(live: false)
  end

  it 'inits hash' do
    @mid.hash.nil?.should == false
  end

  it 'has demo service url' do
    @mid.url.should == 'https://tsp.demo.sk.ee/mid-api'
  end

  it 'has demo service name' do
    @mid.name.should == 'DEMO'
  end

  it 'has demo service uuid' do
    @mid.uuid.should == '00000000-0000-0000-0000-000000000000'
  end

  it 'gets auth hash with session id' do
    auth = @mid.authenticate!(phone: '00000766', personal_code: '60001019906')
    auth[:session_id].nil?.should
    auth[:phone].should
    auth[:phone_calling_code].should == '+372'
  end

  it 'gets verified user attributes' do
    auth = @mid.authenticate!(phone: '00000766', personal_code: '60001019906')
    verify = @mid.verify!(auth)
    verify.should ==
      {
        'personal_code' => '60001019906',
        'first_name' => 'MARY ÄNN',
        'last_name' => 'O’CONNEŽ-ŠUSLIK TESTNUMBER',
        'phone' => '00000766',
        'phone_calling_code' => '+372',
        'auth_provider' => 'mobileid'
      }
  end

  it 'raises error with response code' do
    lambda {
      @mid.long_poll!(session_id: 'wrongid', doc: '')
    }.should raise_error(MobileId::Error, /There was some error 400/)
  end

  it 'calculates verification code' do
    @mid = MobileId::Auth.new(live: false)

    @mid.init_doc('test')
    @mid.verification_code.should == '5000'
  end
end


lib/mobile_id/cert.rb
# frozen_string_literal: true

module MobileId
  class Cert
    class << self
      def root_path
        @root_path ||= File.expand_path('certs', __dir__)
      end

      def live_store
        @live_store ||=
          build_store(
            [
              File.join(root_path, 'EE_Certification_Centre_Root_CA.pem.crt'),
              File.join(root_path, 'EE-GovCA2018.pem.crt'),
              File.join(root_path, 'EID-SK_2011.pem.crt'),
              File.join(root_path, 'EID-SK_2016.pem.crt'),
              File.join(root_path, 'esteid2018.pem.crt'),
              File.join(root_path, 'ESTEID-SK_2011.pem.crt'),
              File.join(root_path, 'ESTEID-SK_2015.pem.crt'),
              File.join(root_path, 'KLASS3-SK_2010_EECCRCA.pem.crt'),
              File.join(root_path, 'KLASS3-SK_2010_EECCRCA_SHA384.pem.crt'),
              File.join(root_path, 'KLASS3-SK_2016_EECCRCA_SHA384.pem.crt'),
              File.join(root_path, 'KLASS3-SK.pem.crt'),
              File.join(root_path, 'NQ-SK_2016.pem.crt')
            ]
          )
      end

      def test_store
        @test_store ||=
          build_store(
            [
              File.join(root_path, 'TEST_of_EE_Certification_Centre_Root_CA.pem.crt'),
              File.join(root_path, 'TEST_of_ESTEID-SK_2015.pem.crt')
            ]
          )
      end

      def build_store(paths)
        store = OpenSSL::X509::Store.new
        paths.each do |path|
          cert = OpenSSL::X509::Certificate.new(File.read(path))
          store.add_cert(cert)
        end
        store
      end
    end

    attr_accessor :cert, :subject

    def initialize(base64_cert, live:)
      self.cert = OpenSSL::X509::Certificate.new(Base64.decode64(base64_cert))
      verify!(cert, live: live)
      build_cert_subject
    end

    def verify!(cert, live:)
      if live == true
        raise Error, 'User certificate is not valid' unless self.class.live_store.verify(cert)
      else
        unless self.class.test_store.verify(cert) || self.class.live_store.verify(cert)
          raise Error,
                'User certificate is not valid'
        end
      end

      raise Error, 'User certificate is not valid [check_key]' unless cert.public_key.check_key
      raise Error, 'User certificate is expired' unless (cert.not_before...cert.not_after).include?(Time.now)

      true
    end

    def verify_signature!(signature_base64, doc)
      signature = Base64.decode64(signature_base64)
      digest = OpenSSL::Digest::SHA256.new(doc)

      valid =
        begin
          cert.public_key.verify(digest, signature, doc)
        rescue OpenSSL::PKey::PKeyError
          der_signature = cvc_to_der(signature) # Probably signature is CVC encoded
          cert.public_key.verify(digest, der_signature, doc)
        end

      raise Error, 'We could not verify user signature' unless valid
    end

    def cvc_to_der(cvc)
      sign_hex = cvc.unpack1('H*')
      half = sign_hex.size / 2
      i = [OpenSSL::ASN1::Integer.new(sign_hex[0...half].to_i(16)),
           OpenSSL::ASN1::Integer.new(sign_hex[half..sign_hex.size].to_i(16))]
      seq = OpenSSL::ASN1::Sequence.new(i)
      seq.to_der
    end

    def given_name
      subject['GN'].tr(',', ' ')
    end
    alias first_name given_name

    def surname
      subject['SN'].tr(',', ' ')
    end
    alias last_name surname

    def country
      subject['C'].tr(',', ' ')
    end

    def common_name
      subject['CN']
    end

    def organizational_unit
      subject['OU']
    end

    def serial_number
      subject['serialNumber']
    end
    alias personal_code serial_number

    private

    def build_cert_subject
      self.subject = cert.subject.to_utf8.split(/(?<!\\),+/).each_with_object({}) do |c, result|
        next unless c.include?('=')

        key, val = c.split('=')
        result[key] = val
      end
    end
  end
end


lib/mobile_id/auth.rb
# frozen_string_literal: true

module MobileId
  class Auth
    # API documentation https://github.com/SK-EID/MID
    LIVE_URL = 'https://mid.sk.ee/mid-api'
    TEST_URL = 'https://tsp.demo.sk.ee/mid-api'

    TEST_UUID  = '00000000-0000-0000-0000-000000000000'
    TEST_NAME  = 'DEMO'

    attr_accessor :url, :uuid, :name, :doc, :hash, :user_cert, :live

    def initialize(live:, uuid: nil, name: nil, doc: nil)
      self.url = live == true ? LIVE_URL : TEST_URL
      self.uuid = live == true ? uuid : TEST_UUID
      self.name = live == true ? name : TEST_NAME
      self.live = live
      init_doc(doc || SecureRandom.hex(40))
    end

    def init_doc(doc)
      self.doc = doc
      self.hash = Digest::SHA256.digest(doc)
    end

    def authenticate!(phone:, personal_code:, phone_calling_code: nil, language: nil, display_text: nil)
      phone_calling_code ||= '+372'
      full_phone = "#{phone_calling_code}#{phone}"
      language ||=
        case I18n.locale
        when :et
          display_text ||= 'Autentimine'
          'EST'
        when :ru
          display_text ||= 'Аутентификация'
          'RUS'
        else
          display_text ||= 'Authentication'
          'ENG'
        end

      options = {
        headers: {
          'Content-Type': 'application/json'
        },
        query: {},
        body: {
          relyingPartyUUID: uuid,
          relyingPartyName: name,
          phoneNumber: full_phone.to_s.strip,
          nationalIdentityNumber: personal_code.to_s.strip,
          hash: Base64.strict_encode64(hash),
          hashType: 'SHA256',
          language: language,
          displayText: display_text,
          displayTextFormat: 'GSM-7' # or "UCS-2”
        }.to_json
      }

      response = HTTParty.post("#{url}/authentication", options)
      raise Error, "#{I18n.t('mobile_id.some_error')} #{response}" unless response.code == 200

      ActiveSupport::HashWithIndifferentAccess.new(
        session_id: response['sessionID'],
        phone: phone,
        phone_calling_code: phone_calling_code,
        doc: doc
      )
    end

    def verify!(auth)
      long_poll!(session_id: auth['session_id'], doc: auth['doc'])

      ActiveSupport::HashWithIndifferentAccess.new(
        personal_code: personal_code,
        first_name: first_name,
        last_name: last_name,
        phone: auth['phone'],
        phone_calling_code: auth['phone_calling_code'],
        auth_provider: 'mobileid' # User::MOBILEID
      )
    end

    def session_request(session_id)
      response = HTTParty.get(url + "/authentication/session/#{session_id}")
      raise Error, "#{I18n.t('mobile_id.some_error')} #{response.code} #{response}" if response.code != 200

      response
    end

    def long_poll!(session_id:, doc:)
      response = nil

      # Retries until RUNNING state turns to COMPLETE
      30.times do |_i|
        response = session_request(session_id)
        break if response['state'] == 'COMPLETE'

        sleep 1
      end
      raise Error, "#{I18n.t('mobile_id.some_error')} #{response.code} #{response}" if response['state'] != 'COMPLETE'

      if response['result'] != 'OK'
        message =
          case response['result']
          when 'TIMEOUT'
            I18n.t('mobile_id.timeout')
          when 'NOT_MID_CLIENT'
            I18n.t('mobile_id.user_is_not_mobile_id_client')
          when 'USER_CANCELLED'
            I18n.t('mobile_id.user_cancelled')
          when 'SIGNATURE_HASH_MISMATCH'
            I18n.t('mobile_id.signature_hash_mismatch')
          when 'PHONE_ABSENT'
            I18n.t('mobile_id.phone_absent')
          when 'DELIVERY_ERROR'
            I18n.t('mobile_id.delivery_error')
          when 'SIM_ERROR'
            I18n.t('mobile_id.sim_error')
          end
        raise Error, message
      end

      @user_cert = MobileId::Cert.new(response['cert'], live: live)
      @user_cert.verify_signature!(response['signature']['value'], doc)
      self.user_cert = @user_cert
    end

    def verification_code
      binary = hash.to_s.unpack1('B*')
      '%04d' % (binary[0...6] + binary[-7..]).to_i(2)
    end

    def given_name
      user_cert.given_name
    end
    alias first_name given_name

    def surname
      user_cert.surname
    end
    alias last_name surname

    def country
      user_cert.country
    end

    def common_name
      user_cert.common_name
    end

    def organizational_unit
      user_cert.organizational_unit
    end

    def serial_number
      user_cert.serial_number
    end
    alias personal_code serial_number
  end
end
